File: dotnet/src/SimpleOrm.Sqlite/SqliteDialect.cs
using System.Data.Common;
using System.Text;
using Microsoft.Data.Sqlite;

namespace SimpleOrm.Sqlite;

/// <summary>SQLite implementation of <see cref="IDialect"/> backed by Microsoft.Data.Sqlite.</summary>
public sealed class SqliteDialect : IDialect
{
    public DbConnection CreateConnection(string connectionString)
        => new SqliteConnection(connectionString);

    /// <summary>Unquoted (ADR-0024): snake_case names need nothing, and the reference renderings stay byte-identical.</summary>
    public string QuoteIdentifier(string identifier) => identifier;

    public string LimitOffsetClause(string? limitParameter, string? offsetParameter) => (limitParameter, offsetParameter) switch
    {
        (not null, not null) => $"limit {limitParameter} offset {offsetParameter}",
        (not null, null) => $"limit {limitParameter}",
        (null, not null) => $"limit -1 offset {offsetParameter}",   // SQLite requires LIMIT before OFFSET
        _ => string.Empty,
    };

    public string SelectSql(SelectAst select, BindCriteriaParameter bindParameter)
        => AnsiSelectRenderer.SelectSql(this, select, bindParameter);

    public bool SupportsArrayParameters => false;

    public bool BindsTemporalsNatively => false;   // TEXT storage: the ISO-8601 string IS the value (§7.9)

    public bool PagingRequiresOrderBy => false;

    public bool SupportsRowValueIn => true;

    public bool SupportsMaterializedViews => false;

    public bool SupportsProcedures => false;

    public bool SupportsTransactionalDdl => true;

    public string ColumnsInfoSql
        => "select name, type, \"notnull\", pk from pragma_table_info(@relation)";

    public string ViewDefinitionSql
        => "select sql from sqlite_master where type = 'view' and name = @relation";

    public string IndexesInfoSql
        => "select il.name, il.\"unique\", ii.seqno, ii.name, ii.\"desc\" "
            + "from pragma_index_list(@relation) il, pragma_index_xinfo(il.name) ii "
            + "where il.origin = 'c' and ii.key = 1 order by il.name, ii.seqno";

    public bool IsDeclaredTypeCompatible(string declaredType, Type clrType, bool enumAsInt)
    {
        var type = Nullable.GetUnderlyingType(clrType) ?? clrType;
        var declared = declaredType.Trim().ToUpperInvariant();
        if (declared is "" or "ANY")
        {
            return true;   // untyped: SQLite enforces nothing, nothing to contradict
        }

        if (type.IsEnum)
        {
            return enumAsInt ? declared is "INT" or "INTEGER" : declared == "TEXT";
        }

        return declared switch
        {
            "INT" or "INTEGER" => type == typeof(int) || type == typeof(long) || type == typeof(short) || type == typeof(bool),
            "REAL" => type == typeof(double) || type == typeof(float),
            "BLOB" => type == typeof(byte[]) || type == typeof(Guid),
            "TEXT" => type == typeof(string) || type == typeof(decimal) || type == typeof(Guid)
                || type == typeof(DateTime) || type == typeof(DateTimeOffset)
                || type.FullName is "System.DateOnly" or "System.TimeOnly",
            _ => false,
        };
    }

    public DbTransaction BeginMigrationRunLock(DbConnection connection)
        => ((SqliteConnection)connection).BeginTransaction(deferred: false);   // BEGIN IMMEDIATE

    public string VersionTableSql
        => """
           create table if not exists schema_version (
               version      INTEGER NOT NULL,
               object       TEXT NOT NULL,
               description  TEXT NOT NULL,
               checksum     TEXT NOT NULL,
               applied_at   TEXT NOT NULL,
               execution_ms INTEGER NOT NULL,
               primary key (version, object)
           ) STRICT
           """;

    // Migration-action DDL (ADR-0024): these strings are frozen — recorded
    // checksums hash them.

    public string RenameTableSql(string fromName, string toName)
        => $"alter table {fromName} rename to {toName}";

    public string RenameColumnSql(string table, string fromName, string toName)
        => $"alter table {table} rename column {fromName} to {toName}";

    public string AddColumnSql(string table, string column, string storageType, bool nullable, string? defaultSql)
        => $"alter table {table} add column {column} {storageType}"
            + (nullable ? string.Empty : " not null")
            + (defaultSql is null ? string.Empty : " default " + defaultSql);

    public string DropColumnSql(string table, string column)
        => $"alter table {table} drop column {column}";

    public string DropTableSql(string table)
        => "drop table " + table;

    public string DropIndexSql(string table, string index)
        => "drop index " + index;

    public string CreateViewSql(EntityMap map)
        => "create view if not exists " + map.RelationName + " as\n" + map.DefiningSql;

    public string CreateTableSql(EntityMap map)
    {
        var builder = new StringBuilder("create table if not exists ")
            .Append(map.RelationName).Append(" (");

        var first = true;
        foreach (var property in map.Properties)
        {
            builder.Append(first ? "\n    " : ",\n    ").Append(property.ColumnName).Append(' ');
            first = false;

            if (map.KeyStrategy == KeyStrategy.DatabaseGenerated && property.IsKey)
            {
                // The exact spelling that makes the column the rowid alias.
                builder.Append("INTEGER PRIMARY KEY");
                continue;
            }

            builder.Append(ColumnType(property));
            if (!property.IsNullable)
            {
                builder.Append(" NOT NULL");
            }

            if (map.KeyStrategy == KeyStrategy.ClientGuid && property.IsKey)
            {
                builder.Append(" PRIMARY KEY");
            }
        }

        if (map.KeyStrategy == KeyStrategy.Natural && map.KeyProperties.Count > 0)
        {
            builder.Append(",\n    primary key (")
                .Append(string.Join(", ", map.KeyProperties.Select(k => k.ColumnName)))
                .Append(')');
        }

        return builder.Append("\n) STRICT").ToString();
    }

    public IReadOnlyList<string> CreateIndexSql(EntityMap map)
        => map.Indexes
            .Select(index =>
                "create " + (index.Unique ? "unique " : string.Empty) + "index if not exists " + index.Name
                + " on " + map.RelationName + " ("
                + string.Join(", ", index.Columns.Select(c => c.ColumnName + (c.Descending ? " desc" : string.Empty)))
                + ")")
            .ToArray();

    public string InsertSql(EntityMap map)
    {
        var columns = map.Properties.Where(p => !p.IsGenerated).ToArray();
        var sql = "insert into " + map.RelationName
            + " (" + string.Join(", ", columns.Select(c => c.ColumnName)) + ")"
            + " values (" + string.Join(", ", columns.Select(c => "@" + c.ColumnName)) + ")";

        if (map.KeyStrategy == KeyStrategy.DatabaseGenerated)
        {
            sql += " returning " + map.KeyProperties[0].ColumnName;
        }

        return sql;
    }

    public string UpdateSql(EntityMap map)
    {
        // Generated non-key columns are database-owned: never in SET (mirrors the
        // insert exclusion and Db.UpdateAsync's binding filter).
        var assignments = map.Properties
            .Where(p => !p.IsKey && !p.IsVersion && !p.IsGenerated)
            .Select(p => p.ColumnName + " = @" + p.ColumnName)
            .ToList();
        if (map.VersionProperty is { } version)
        {
            assignments.Add(version.ColumnName + " = " + version.ColumnName + " + 1");
        }

        return "update " + map.RelationName
            + " set " + string.Join(", ", assignments)
            + " where " + KeyPredicate(map)
            + (map.VersionProperty is { } v ? " and " + v.ColumnName + " = @" + v.ColumnName : string.Empty);
    }

    public string DeleteSql(EntityMap map, bool checkVersion)
        => "delete from " + map.RelationName
            + " where " + KeyPredicate(map)
            + (checkVersion && map.VersionProperty is { } v ? " and " + v.ColumnName + " = @" + v.ColumnName : string.Empty);

    private static string KeyPredicate(EntityMap map)
        => string.Join(" and ", map.KeyProperties.Select(k => k.ColumnName + " = @" + k.ColumnName));

    public string StorageType(PropertyMap property) => ColumnType(property);

    /// <summary>CLR → SQLite storage type per the §7.9 conventions (dates/decimals/GUIDs as TEXT).</summary>
    private static string ColumnType(PropertyMap property)
    {
        var type = Nullable.GetUnderlyingType(property.ClrType) ?? property.ClrType;
        if (type.IsEnum)
        {
            return property.EnumAsInt ? "INTEGER" : "TEXT";
        }

        if (type == typeof(int) || type == typeof(long) || type == typeof(short) || type == typeof(bool))
        {
            return "INTEGER";
        }

        if (type == typeof(double) || type == typeof(float))
        {
            return "REAL";
        }

        if (type == typeof(byte[]))
        {
            return "BLOB";
        }

        // string, decimal, DateTime/Offset, DateOnly/TimeOnly, Guid, JSON, handler types.
        return "TEXT";
    }
}

File: php/src/Dialect/SqliteDialect.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Dialect;

use PDO;
use SimpleOrm\Metadata\ColumnType;
use SimpleOrm\Metadata\EntityIndex;
use SimpleOrm\Metadata\EntityMap;
use SimpleOrm\Metadata\IndexColumn;
use SimpleOrm\Metadata\KeyStrategy;
use SimpleOrm\Metadata\PropertyMap;
use SimpleOrm\Query\AnsiSelectRenderer;
use SimpleOrm\Query\SelectAst;

/**
 * SQLite implementation of {@see Dialect} backed by PDO's `pdo_sqlite` driver
 * (ADR-0003). The reference dialect (CLAUDE.md §7.25): every rendered DDL/CRUD/
 * action string below is byte-identical to `dotnet/src/SimpleOrm.Sqlite/SqliteDialect.cs`
 * — checksums and conformance pins depend on it.
 */
final class SqliteDialect implements Dialect
{
    public function createConnection(string $connectionString): PDO
    {
        $pdo = new PDO('sqlite:' . self::resolvePath($connectionString));
        $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        $pdo->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);
        $pdo->setAttribute(PDO::ATTR_STRINGIFY_FETCHES, false);
        $pdo->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);

        return $pdo;
    }

    /** Accepts a bare file path, a `sqlite:<path>` PDO DSN, or C#-style `Data Source=<path>`. */
    private static function resolvePath(string $connectionString): string
    {
        if (stripos($connectionString, 'sqlite:') === 0) {
            return substr($connectionString, strlen('sqlite:'));
        }

        foreach (explode(';', $connectionString) as $part) {
            $part = trim($part);
            if (stripos($part, 'Data Source=') === 0) {
                return trim(substr($part, strlen('Data Source=')));
            }
        }

        return $connectionString;
    }

    /** Unquoted (ADR-0024): snake_case names need nothing, and the reference renderings stay byte-identical. */
    public function quoteIdentifier(string $identifier): string
    {
        return $identifier;
    }

    public function createTableSql(EntityMap $map): string
    {
        $lines = [];
        foreach ($map->properties as $property) {
            if ($map->keyStrategy === KeyStrategy::DatabaseGenerated && $property->key) {
                // The exact spelling that makes the column the rowid alias.
                $lines[] = $property->columnName . ' INTEGER PRIMARY KEY';
                continue;
            }

            $line = $property->columnName . ' ' . self::columnType($property);
            if (!$property->nullable) {
                $line .= ' NOT NULL';
            }

            if ($map->keyStrategy === KeyStrategy::ClientGuid && $property->key) {
                $line .= ' PRIMARY KEY';
            }

            $lines[] = $line;
        }

        if ($map->keyStrategy === KeyStrategy::Natural && $map->keyProperties !== []) {
            $lines[] = 'primary key (' . implode(', ', array_map(
                static fn (PropertyMap $k): string => $k->columnName,
                $map->keyProperties,
            )) . ')';
        }

        return "create table if not exists {$map->relationName} (\n    " . implode(",\n    ", $lines) . "\n) STRICT";
    }

    /** @return list<string> */
    public function createIndexSql(EntityMap $map): array
    {
        return array_map(
            static fn (EntityIndex $index): string => 'create ' . ($index->unique ? 'unique ' : '') . 'index if not exists '
                . $index->name . ' on ' . $map->relationName . ' (' . implode(', ', array_map(
                    static fn (IndexColumn $c): string => $c->columnName . ($c->descending ? ' desc' : ''),
                    $index->columns,
                )) . ')',
            $map->indexes,
        );
    }

    public function createViewSql(EntityMap $map): string
    {
        return 'create view if not exists ' . $map->relationName . " as\n" . $map->definingSql;
    }

    public function insertSql(EntityMap $map): string
    {
        $columns = array_values(array_filter($map->properties, static fn (PropertyMap $p): bool => !$p->generated));
        $sql = 'insert into ' . $map->relationName
            . ' (' . implode(', ', array_map(static fn (PropertyMap $c): string => $c->columnName, $columns)) . ')'
            . ' values (' . implode(', ', array_map(static fn (PropertyMap $c): string => '@' . $c->columnName, $columns)) . ')';

        if ($map->keyStrategy === KeyStrategy::DatabaseGenerated) {
            $sql .= ' returning ' . $map->keyProperties[0]->columnName;
        }

        return $sql;
    }

    public function updateSql(EntityMap $map): string
    {
        // Generated non-key columns are database-owned: never in SET (mirrors
        // the insert exclusion).
        return self::renderUpdate($map, array_values(array_filter(
            $map->properties,
            static fn (PropertyMap $p): bool => !$p->key && !$p->version && !$p->generated,
        )));
    }

    public function updateOnlySql(EntityMap $map, array $properties): string
    {
        return self::renderUpdate($map, $properties);
    }

    /** @param list<PropertyMap> $set */
    private static function renderUpdate(EntityMap $map, array $set): string
    {
        $assignments = array_map(
            static fn (PropertyMap $p): string => $p->columnName . ' = @' . $p->columnName,
            $set,
        );
        if ($map->versionProperty !== null) {
            $assignments[] = $map->versionProperty->columnName . ' = ' . $map->versionProperty->columnName . ' + 1';
        }

        $sql = 'update ' . $map->relationName
            . ' set ' . implode(', ', $assignments)
            . ' where ' . self::keyPredicate($map);
        if ($map->versionProperty !== null) {
            $sql .= ' and ' . $map->versionProperty->columnName . ' = @' . $map->versionProperty->columnName;
        }

        return $sql;
    }

    public function deleteSql(EntityMap $map, bool $checkVersion): string
    {
        $sql = 'delete from ' . $map->relationName . ' where ' . self::keyPredicate($map);
        if ($checkVersion && $map->versionProperty !== null) {
            $sql .= ' and ' . $map->versionProperty->columnName . ' = @' . $map->versionProperty->columnName;
        }

        return $sql;
    }

    private static function keyPredicate(EntityMap $map): string
    {
        return implode(' and ', array_map(
            static fn (PropertyMap $k): string => $k->columnName . ' = @' . $k->columnName,
            $map->keyProperties,
        ));
    }

    public function limitOffsetClause(?string $limitParameter, ?string $offsetParameter): string
    {
        return match (true) {
            $limitParameter !== null && $offsetParameter !== null => "limit {$limitParameter} offset {$offsetParameter}",
            $limitParameter !== null => "limit {$limitParameter}",
            // SQLite requires LIMIT before OFFSET.
            $offsetParameter !== null => "limit -1 offset {$offsetParameter}",
            default => '',
        };
    }

    public function selectSql(SelectAst $select, callable $bindParameter): string
    {
        return AnsiSelectRenderer::selectSql($this, $select, $bindParameter);
    }

    public function pagingRequiresOrderBy(): bool
    {
        return false;
    }

    public function supportsRowValueIn(): bool
    {
        return true;
    }

    public function supportsArrayParameters(): bool
    {
        return false;
    }

    public function bindsTemporalsNatively(): bool
    {
        return false;   // TEXT storage: the ISO-8601 string IS the value (§7.9)
    }

    public function supportsMaterializedViews(): bool
    {
        return false;
    }

    public function supportsProcedures(): bool
    {
        return false;
    }

    public function supportsTransactionalDdl(): bool
    {
        return true;
    }

    public function columnsInfoSql(): string
    {
        return 'select name, type, "notnull", pk from pragma_table_info(@relation)';
    }

    public function viewDefinitionSql(): string
    {
        return "select sql from sqlite_master where type = 'view' and name = @relation";
    }

    public function indexesInfoSql(): string
    {
        return 'select il.name, il."unique", ii.seqno, ii.name, ii."desc" '
            . 'from pragma_index_list(@relation) il, pragma_index_xinfo(il.name) ii '
            . "where il.origin = 'c' and ii.key = 1 order by il.name, ii.seqno";
    }

    public function isDeclaredTypeCompatible(string $declaredType, ColumnType $type): bool
    {
        $declared = strtoupper(trim($declaredType));
        if ($declared === '' || $declared === 'ANY') {
            return true;   // untyped: SQLite enforces nothing, nothing to contradict
        }

        return match ($declared) {
            'INT', 'INTEGER' => in_array(
                $type,
                [ColumnType::Int16, ColumnType::Int32, ColumnType::Int64, ColumnType::Bool, ColumnType::EnumInt],
                true,
            ),
            'REAL' => $type === ColumnType::Double || $type === ColumnType::Float,
            'BLOB' => $type === ColumnType::Bytes || $type === ColumnType::Guid,
            'TEXT' => in_array(
                $type,
                [
                    ColumnType::String, ColumnType::Decimal, ColumnType::Guid, ColumnType::DateTime,
                    ColumnType::DateTimeOffset, ColumnType::Date, ColumnType::Time, ColumnType::EnumText,
                ],
                true,
            ),
            default => false,
        };
    }

    public function storageType(PropertyMap $property): string
    {
        return self::columnType($property);
    }

    /** CLR → SQLite storage type per the §7.9 conventions (dates/decimals/GUIDs as TEXT). */
    private static function columnType(PropertyMap $property): string
    {
        return match ($property->type) {
            ColumnType::Int16, ColumnType::Int32, ColumnType::Int64, ColumnType::Bool, ColumnType::EnumInt => 'INTEGER',
            ColumnType::Double, ColumnType::Float => 'REAL',
            ColumnType::Bytes => 'BLOB',
            default => 'TEXT',   // string, decimal, DateTime/Offset, date, time, guid, enum_text, custom
        };
    }

    public function beginMigrationRunLock(PDO $connection): void
    {
        $connection->exec('BEGIN IMMEDIATE');
    }

    public function versionTableSql(): string
    {
        return <<<'SQL'
            create table if not exists schema_version (
                version      INTEGER NOT NULL,
                object       TEXT NOT NULL,
                description  TEXT NOT NULL,
                checksum     TEXT NOT NULL,
                applied_at   TEXT NOT NULL,
                execution_ms INTEGER NOT NULL,
                primary key (version, object)
            ) STRICT
            SQL;
    }

    // Migration-action DDL (ADR-0024): these strings are frozen — recorded checksums hash them.

    public function renameTableSql(string $fromName, string $toName): string
    {
        return "alter table {$fromName} rename to {$toName}";
    }

    public function renameColumnSql(string $table, string $fromName, string $toName): string
    {
        return "alter table {$table} rename column {$fromName} to {$toName}";
    }

    public function addColumnSql(string $table, string $column, string $storageType, bool $nullable, ?string $defaultSql): string
    {
        return "alter table {$table} add column {$column} {$storageType}"
            . ($nullable ? '' : ' not null')
            . ($defaultSql === null ? '' : ' default ' . $defaultSql);
    }

    public function dropColumnSql(string $table, string $column): string
    {
        return "alter table {$table} drop column {$column}";
    }

    public function dropTableSql(string $table): string
    {
        return 'drop table ' . $table;
    }

    public function dropIndexSql(string $table, string $index): string
    {
        return 'drop index ' . $index;
    }
}
